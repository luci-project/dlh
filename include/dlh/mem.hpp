// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#pragma once

#include <dlh/types.hpp>
#include <dlh/math.hpp>

namespace Memory {
/*! \brief Allocate a memory block
 * The memory is not initialized.
 *
 * \param size Requested size of memory in bytes.
 * \return Address of memory or `0` on error (no memory available) or if size
 *         was zero.
 */
uintptr_t alloc(size_t size);

template<typename T>
inline T* alloc(size_t size) {
	return reinterpret_cast<T*>(alloc(size));
}

/*! \brief Free an allocated memory block
 *
 * \param addr Address of an previously allocated memory block.
 */
void free(uintptr_t addr);

template<typename T>
inline void free(T* addr) {
	return free(reinterpret_cast<uintptr_t>(addr));
}


/*! \brief Change the size of an allocated memory block
 * The contents will be unchanged in the range from the start of the region up
 * to the minimum of the old and new sizes. If the new size is larger than the
 * old size, the added memory will not be initialized.
 *
 * \param addr Address of an previously allocated memory block.
 *             If `0`, then the call is equivalent to \ref alloc().
 * \param size New size of the memory block. If equal to zero, then the call
 *             is equivalent to free
 * \return Address of new memory block or `0` on error
 */
uintptr_t realloc(uintptr_t addr, size_t size);

template<typename T>
inline T* realloc(T* addr, size_t size) {
	return reinterpret_cast<T*>(realloc(reinterpret_cast<uintptr_t>(addr), size));
}

/*! \brief Allocate memory for an array of elements
 * The memory is set to zero.
 *
 * \param nmemb Number of elements
 * \param size Size of an element in bytes
 * \return address of the allocated memory or `0` if the request fails
 */
uintptr_t alloc_array(size_t nmemb, size_t size);

template<typename T>
inline T* alloc_array(size_t nmemb) {
	return reinterpret_cast<T*>(alloc_array(nmemb, sizeof(T)));
}

/*! \brief Copy a memory area
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return Address of destination
 * \note The memory must not overlap!
 */
inline uintptr_t copy(uintptr_t dest, uintptr_t src, size_t size) {
	unsigned char * __restrict__ destination = reinterpret_cast<unsigned char*>(dest);
	unsigned char const * __restrict__ source = reinterpret_cast<unsigned char const*>(src);

	for(size_t i = 0; i != size; ++i) {
		destination[i] = source[i];
	}

	return dest;
}

/*! \brief Copy a memory area
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return pointer to destination
 * \note The memory must not overlap!
 */
template<typename T, typename U>
inline T* copy(T * __restrict__ dest, U const * __restrict__ src, size_t size) {
	return reinterpret_cast<T*>(copy(reinterpret_cast<uintptr_t>(dest), reinterpret_cast<uintptr_t>(src), size));
}

/*! \brief Copy a memory area
 * while the source may overlap with the destination
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return Address of destination
 */
inline uintptr_t move(uintptr_t dest, uintptr_t src, size_t size) {
	unsigned char * __restrict__ destination = reinterpret_cast<unsigned char*>(dest);
	unsigned char const * __restrict__ source = reinterpret_cast<unsigned char const*>(src);

	if(source > destination) {
		for(size_t i = 0; i != size; ++i)
			destination[i] = source[i];
	} else {
		for(size_t i = size; i != 0; --i)
			destination[i-1] = source[i-1];
	}

	return dest;
}

/*! \brief Copy a memory area
 * while the source may overlap with the destination
 * \ingroup string
 * \param dest destination buffer
 * \param src source buffer
 * \param size number of bytes to copy
 * \return pointer to destination
 */
template<typename T, typename U>
inline T* move(T * dest, U const * src, size_t size) {
	return reinterpret_cast<T*>(move(reinterpret_cast<uintptr_t>(dest), reinterpret_cast<uintptr_t>(src), size));
}

/*! \brief Fill a memory area with a pattern
 * \ingroup string
 * \param dest destination buffer
 * \param pattern single byte pattern
 * \param size number of bytes to fill with pattern
 * \return address of destination
 */
inline uintptr_t set(uintptr_t dest, int pattern, size_t size) {
	unsigned char *destination = reinterpret_cast<unsigned char*>(dest);

	for(size_t i = 0; i != size; ++i) {
		destination[i] = static_cast<unsigned char>(pattern);
	}

	return dest;
}

/*! \brief Fill a memory area with a pattern
 * \ingroup string
 * \param dest destination buffer
 * \param pattern single byte pattern
 * \param size number of bytes to fill with pattern
 * \return pointer to destination
 */
template<typename T>
inline T* set(T* dest, int pattern, size_t size) {
	return reinterpret_cast<T*>(set(reinterpret_cast<uintptr_t>(dest), pattern, size));
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 pointer to first element
 * \param s2 pointer to second element
 * \param size number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
inline int compare(uintptr_t s1, uintptr_t s2, size_t n) {
	const unsigned char * c1 = reinterpret_cast<const unsigned char*>(s1);
	const unsigned char * c2 = reinterpret_cast<const unsigned char*>(s2);

	for(size_t i = 0; i != n; ++i)
		if (c1[i] != c2[i])
			return static_cast<int>(c1[i]) - static_cast<int>(c2[i]);

	return 0;
}


/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 pointer to first element
 * \param s2 pointer to second element
 * \param size number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T * s1, const U * s2, size_t n) {
	return compare(reinterpret_cast<uintptr_t>(s1), reinterpret_cast<uintptr_t>(s2), n);
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 pointer to first element
 * \param s2 pointer to second element
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T * s1, const U * s2) {
	int r = compare(s1, s2, Math::min(sizeof(T), sizeof(U)));
	return r == 0 ? sizeof(U) - sizeof(T) : r;
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 reference to first element
 * \param s2 reference to second element
 * \param size number of bytes to compare
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T & s1, const T & s2, size_t n) {
	return compare(reinterpret_cast<uintptr_t>(&s1), reinterpret_cast<uintptr_t>(&s2), n);
}

/*! \brief Compare a memory area
 * \ingroup string
 * \param s1 reference to first element
 * \param s2 reference to second element
 * \return an integer less than, equal to, or greater than zero if the given number of bytes of the first element are
 *          found, respectively, to be less than, to match, or be greater than second element
 */
template<typename T, typename U>
inline int compare(const T & s1, const U & s2) {
	int r = compare(reinterpret_cast<uintptr_t>(&s1), reinterpret_cast<uintptr_t>(&s2), Math::min(sizeof(T), sizeof(U)));
	return r == 0 ? sizeof(U) - sizeof(T) : r;
}

}  // namespace Memory
