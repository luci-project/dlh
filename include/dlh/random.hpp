// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#pragma once

#include "types.hpp"

/*! Mersenne Twister (32 bit pseudorandom number generator)
 *
 * \see [Wikipedia](https://en.wikipedia.org/wiki/Mersenne_Twister)
 */
class Random {
 private:
	static const uint32_t N = 624;
	static const uint32_t M = 397;
	static const uint32_t R = 31;
	static const uint32_t A = 0x9908B0DF;

	// initialization multiplier
	static const uint32_t F = 1812433253;

	static const uint32_t U = 11;

	static const uint32_t S = 7;
	static const uint32_t B = 0x9D2C5680;

	static const uint32_t T = 15;
	static const uint32_t C = 0xEFC60000;

	static const uint32_t L = 18;

	static const uint32_t MASK_LOWER = (1UL << R) - 1;
	static const uint32_t MASK_UPPER = (1UL << R);

	uint32_t mt[N];
	uint16_t index;

 public:
	/*! \brief Constructor
	 *
	 *  \param seed initial value (seed) for the pseudorandom number generator.
	 */
	explicit Random(uint32_t seed = 0) {
		set(seed);
	}

	/*! \brief (Re)Seed
	 *
	 *  \param seed initial value (seed) for the pseudorandom number generator.
	 */
	void set(uint32_t seed) {
		mt[0] = seed;

		for (uint32_t i = 1; i < N; i++) {
			mt[i] = (F * (mt[i - 1] ^ (mt[i - 1] >> 30)) + i);
		}

		index = N;
	}

	/*! \brief Get the next (pseudo)random number.
	 *  \return some number
	 */
	uint32_t number() {
		uint16_t v = index;

		if (index >= N) {
			// Twist
			for (uint32_t i = 0; i < N; i++) {
				uint32_t x = (mt[i] & MASK_UPPER) + (mt[(i + 1) % N] & MASK_LOWER);

				uint32_t xA = x >> 1;

				if ((x & 0x1) != 0) {
					xA ^= A;
				}

				mt[i] = mt[(i + M) % N] ^ xA;
			}
			v = index = 0;
		}

		uint32_t y = mt[v];
		index = v + 1;

		y ^= (y >> U);
		y ^= (y << S) & B;
		y ^= (y << T) & C;
		y ^= (y >> L);

		return y;
	}
};
