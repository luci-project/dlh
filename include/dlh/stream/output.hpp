#pragma once

#include <dlh/syscall.hpp>
#include <dlh/stream/string.hpp>

template<size_t BUFFERSZ>
class OutputStream : public StringStream<BUFFERSZ> {
 protected:
	int fd;

 public:
	/*! \brief Default constructor  */
	explicit OutputStream(int fd) : fd(fd) {}

	/*! \brief Destructor */
	virtual ~OutputStream() {}

	/*! \brief Flush the buffer.
	 */
	virtual void flush() override {
		size_t l = 0;
		while (l < this->_pos) {
			auto r = Syscall::write(fd, this->_bufptr + l, this->_pos - l);
			if (!r.valid())
				break;
			else
				l += r.value();
		}
		this->_pos = 0;
	}

};

extern OutputStream<1024> cout, cerr;
