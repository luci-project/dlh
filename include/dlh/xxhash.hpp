// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// Copyright 2016 by Stephan Brumme (XXHash)
// Copyright 2014 by Yann Collet (XXHash)
// SPDX-License-Identifier: AGPL-3.0-or-later

/*! \file XXHash (64 bit), based on Yann Collet's descriptions
 * \see http://cyan4973.github.io/xxHash/
 *
 * Copyright (c) 2016 Stephan Brumme. All rights reserved.
 * \see http://create.stephan-brumme.com/disclaimer.html
 *
 * Modified by Bernhard Heinloth, 2021
 */

#pragma once

#include <dlh/types.hpp>
#include <dlh/bytebuffer.hpp>

class XXHash64 {
	/// magic constants :-)
	static const uint64_t Prime1 = 11400714785074694791ULL;
	static const uint64_t Prime2 = 14029467366897019727ULL;
	static const uint64_t Prime3 =  1609587929392839161ULL;
	static const uint64_t Prime4 =  9650029242287828579ULL;
	static const uint64_t Prime5 =  2870177450012600261ULL;

	/// temporarily store up to 31 bytes between multiple add() calls
	static const uint64_t MaxBufferSize = 31+1;

	uint64_t      state[4];
	unsigned char buffer[MaxBufferSize];
	unsigned int  bufferSize;
	uint64_t      totalLength;

	/// rotate bits, should compile to a single CPU instruction (ROL)
	static constexpr uint64_t rotateLeft(uint64_t x, unsigned char bits) {
		return (x << bits) | (x >> (64 - bits));
	}

	/// process a single 64 bit value
	static constexpr uint64_t processSingle(uint64_t previous, uint64_t input) {
		return rotateLeft(previous + input * Prime2, 31) * Prime1;
	}

	/// process a block of 4x4 bytes, this is the main part of the XXHash32 algorithm
	static constexpr void process(const void* data, uint64_t& state0, uint64_t& state1, uint64_t& state2, uint64_t& state3) {
		const uint64_t* block = (const uint64_t*) data;
		state0 = processSingle(state0, block[0]);
		state1 = processSingle(state1, block[1]);
		state2 = processSingle(state2, block[2]);
		state3 = processSingle(state3, block[3]);
	}

 public:
	/*! \brief create new XXHash (64 bit)
	*  \param seed your seed value, even zero is a valid seed
	*/
	explicit constexpr XXHash64(uint64_t seed)
	  : state{seed + Prime1 + Prime2, seed + Prime2, seed, seed - Prime1},
		buffer{0}, bufferSize(0), totalLength(0) {}

	/*! \brief add a chunk of zero bytes
	 * \param length number of zero bytes
	 * \return `false` if length is zero
	 */
	constexpr bool addZeros(uint64_t length) {
		// no data ?
		if (length == 0)
			return false;

		totalLength += length;

		// unprocessed old data plus new data still fit in temporary buffer ?
		if (bufferSize + length < MaxBufferSize) {
			// just add new data
			while (length-- > 0)
				buffer[bufferSize++] = 0;
			return true;
		}

		// copying state to local variables helps optimizer A LOT
		uint64_t s0 = state[0], s1 = state[1], s2 = state[2], s3 = state[3];

		if (bufferSize > 0) {
			// make sure temporary buffer is full (16 bytes)
			length -= MaxBufferSize - bufferSize;
			while (bufferSize < MaxBufferSize)
				buffer[bufferSize++] = 0;

			// process these 32 bytes (4x8)
			process(buffer, s0, s1, s2, s3);
		}

		// 32 bytes at once
		while (length >= MaxBufferSize) {
			// local variables s0..s3 instead of state[0]..state[3] are much faster
			s0 = rotateLeft(s0, 31) * Prime1;
			s1 = rotateLeft(s1, 31) * Prime1;
			s2 = rotateLeft(s2, 31) * Prime1;
			s3 = rotateLeft(s3, 31) * Prime1;
			length -= 32;
		}

		// copy back
		state[0] = s0;
		state[1] = s1;
		state[2] = s2;
		state[3] = s3;

		// copy remainder to temporary buffer
		bufferSize = length;
		for (unsigned int i = 0; i < length; i++)
			buffer[i] = 0;

		// done
		return true;
	}

	/*! \brief add a chunk of bytes
	 * \param input pointer to a continuous block of data (or nullptr to add zeros)
	 * \param length number of bytes
	 * \return `false` if parameters are invalid / zero
	 */
	constexpr bool add(const void* input, uint64_t length) {
		// no data ?
		if (length == 0)
			return false;
		else if (input == nullptr)
			return addZeros(length);

		totalLength += length;
		// byte-wise access
		const unsigned char* data = (const unsigned char*)input;

		// unprocessed old data plus new data still fit in temporary buffer ?
		if (bufferSize + length < MaxBufferSize) {
			// just add new data
			while (length-- > 0)
				buffer[bufferSize++] = *data++;
			return true;
		}

		// point beyond last byte
		const unsigned char* stop = data + length;

		// copying state to local variables helps optimizer A LOT
		uint64_t s0 = state[0], s1 = state[1], s2 = state[2], s3 = state[3];

		// some data left from previous update ?
		if (bufferSize > 0) {
			// make sure temporary buffer is full (16 bytes)
			while (bufferSize < MaxBufferSize)
				buffer[bufferSize++] = *data++;

			// process these 32 bytes (4x8)
			process(buffer, s0, s1, s2, s3);
		}

		// 32 bytes at once
		while (data + MaxBufferSize <= stop) {
			// local variables s0..s3 instead of state[0]..state[3] are much faster
			process(data, s0, s1, s2, s3);
			data += 32;
		}

		// copy back
		state[0] = s0;
		state[1] = s1;
		state[2] = s2;
		state[3] = s3;

		// copy remainder to temporary buffer
		bufferSize = stop - data;
		for (unsigned int i = 0; i < bufferSize; i++)
			buffer[i] = data[i];

		// done
		return true;
	}

	/*! \brief add a chunk of bytes
	 * \param input address of a continuous block of data (or `0` to add zeros)
	 * \param length number of bytes
	 * \return `false` if parameters are invalid / zero
	 */
	constexpr bool add(const uintptr_t input, uint64_t length) {
		return add(reinterpret_cast<const void*>(input), length);
	}

	template<typename T>
	constexpr bool add(T input) {
		return add(&input, sizeof(T));
	}

	template<size_t CAPACITY>
	constexpr bool add(const ByteBuffer<CAPACITY> & bb) {
		return add(bb.buffer(), bb.size());
	}

	/*! \brief get current hash
	 * \return 64 bit XXHash
	 */
	constexpr uint64_t hash() const {
		// fold 256 bit state into one single 64 bit value
		uint64_t result = 0;
		if (totalLength >= MaxBufferSize) {
			result = rotateLeft(state[0],  1) +
			         rotateLeft(state[1],  7) +
			         rotateLeft(state[2], 12) +
			         rotateLeft(state[3], 18);
			result = (result ^ processSingle(0, state[0])) * Prime1 + Prime4;
			result = (result ^ processSingle(0, state[1])) * Prime1 + Prime4;
			result = (result ^ processSingle(0, state[2])) * Prime1 + Prime4;
			result = (result ^ processSingle(0, state[3])) * Prime1 + Prime4;
		} else {
			// internal state wasn't set in add(), therefore original seed is still stored in state2
			result = state[2] + Prime5;
		}

		result += totalLength;

		// process remaining bytes in temporary buffer
		const unsigned char* data = buffer;
		// point beyond last byte
		const unsigned char* stop = data + bufferSize;

		// at least 8 bytes left ? => eat 8 bytes per step
		for (; data + 8 <= stop; data += 8)
			result = rotateLeft(result ^ processSingle(0, *reinterpret_cast<const uint64_t*>(data)), 27) * Prime1 + Prime4;

		// 4 bytes left ? => eat those
		if (data + 4 <= stop) {
			result = rotateLeft(result ^ (*reinterpret_cast<const uint32_t*>(data)) * Prime1,   23) * Prime2 + Prime3;
			data  += 4;
		}

		// take care of remaining 0..3 bytes, eat 1 byte per step
		while (data != stop)
			result = rotateLeft(result ^ (*data++) * Prime5,            11) * Prime1;

		// mix bits
		result ^= result >> 33;
		result *= Prime2;
		result ^= result >> 29;
		result *= Prime3;
		result ^= result >> 32;
		return result;
	}


	/*! \brief combine constructor, add() and hash() in one static function (C style)
	 * \param  input  pointer to a continuous block of data
	 * \param  length number of bytes
	 * \param  seed your seed value, e.g. zero is a valid seed
	 * \return 64 bit XXHash
	 */
	static constexpr uint64_t hash(const void* input, uint64_t length, uint64_t seed) {
		XXHash64 hasher(seed);
		hasher.add(input, length);
		return hasher.hash();
	}
};

template<>
constexpr bool XXHash64::add<const char *>(const char *input) {
	size_t len = 0;
	for (const char * s = input; s != nullptr && *s++ != '\0'; len++) {}
	return add(input, len);
}
