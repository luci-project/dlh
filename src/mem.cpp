// Dirty Little Helper (DLH) - system support library for C/C++
// Copyright 2021-2023 by Bernhard Heinloth <heinloth@cs.fau.de>
// SPDX-License-Identifier: AGPL-3.0-or-later

#include <dlh/mem.hpp>
#include <dlh/log.hpp>
#include <dlh/math.hpp>
#include <dlh/mutex.hpp>
#include <dlh/assert.hpp>
#include <dlh/string.hpp>

#include "alloc_buddy.hpp"

namespace Memory {

#define PAGE_SIZE 0x1000UL

#ifndef MMAP_ALLOC_THREASHOLD
// Threshold to use mmap instead of alloc
#define MMAP_ALLOC_THREASHOLD (32 * PAGE_SIZE)
#endif

#ifndef MMAP_ALLOC_START
// Start address for the mmap base allocator
#define MMAP_ALLOC_START (0x700000000000UL)
#endif

#ifndef MIN_ALLOC_LOG2
// Minimum size to  alloc
// Default: 2^10 = 1024 Bytes
#define MIN_ALLOC_LOG2 10
#endif

#ifndef MAX_ALLOC_LOG2
// Maximum total space to alloc
// Default: 2^30 = 1GB
#define MAX_ALLOC_LOG2 30
#endif

#ifdef DLH_LEGACY
// Use program break based allocator (with 128 KiB chunks)
static uintptr_t reserve_sbrk(size_t size, uintptr_t & max_ptr) {
	const size_t block_size = 128 * 1024;
	size = Math::align_up(size, block_size);


	if (auto sbrk = Syscall::sbrk(size)) {
		uintptr_t ptr = sbrk.value();
		assert(max_ptr == 0 || max_ptr == ptr);
		max_ptr = ptr + size;
		return ptr;
	} else {
		LOG_WARNING << "Allocator reserve by incrementing program break of " << size << " bytes failed: " << sbrk.error_message() << endl;
		return 0;
	}
}
static Allocator::Buddy<MIN_ALLOC_LOG2, MAX_ALLOC_LOG2, reserve_sbrk> allocator;
#else
// Use mmap based allocator
static uintptr_t reserve_mmap(size_t size, uintptr_t & max_ptr) {
	// Mapping address
	static uintptr_t mmap_addr = MMAP_ALLOC_START;

	uintptr_t ptr = 0;

	// Calculate size uising 1 MiB chunks
	const size_t block_size = 256 * PAGE_SIZE;
	size = Math::align_up(size, block_size);

	if (max_ptr == 0) {
		// Initial mapping
		if (auto mmap = Syscall::mmap(mmap_addr, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0)) {
			mmap_addr = ptr = mmap.value();
			max_ptr = ptr + size;
		} else {
			LOG_WARNING << "Allocator reserve by mmap of " << size << " bytes at " << reinterpret_cast<void*>(mmap_addr) << " failed: " << mmap.error_message() << endl;
		}
		return ptr;
	} else {
		// Extend mapping

		/* In case we don't have mremap:
		if (auto mmap = Syscall::mmap(max_ptr, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED_NOREPLACE, 0, 0)) {
			ptr = mmap.value();
			assert(max_ptr == ptr);
			max_ptr += size;
		} else {
			LOG_WARNING << "Allocator reserve by mmap of " << size << " bytes at " << (void*)max_ptr << " failed: " << mmap.error_message() << endl;
		}
		*/

		size_t old_size = max_ptr - mmap_addr;
		size_t new_size = old_size + size;
		assert(old_size % block_size == 0);
		if (auto mremap = Syscall::mremap(mmap_addr, old_size, new_size, 0)) {
			assert(mmap_addr == mremap.value());
			ptr = max_ptr;
			max_ptr += size;
		} else {
			LOG_WARNING << "Allocator reserve by mremap of " << old_size << " bytes to " << new_size << " bytes at " << reinterpret_cast<void*>(mmap_addr) << " failed: " << mremap.error_message() << endl;
		}
	}

	return ptr;
}

static Allocator::Buddy<MIN_ALLOC_LOG2, MAX_ALLOC_LOG2, reserve_mmap> allocator;
#endif

static Mutex mutex;

#ifndef DLH_LEGACY
#define MEMORYMAP_MAGIC 0xDEADBEAFBADF00DUL
class MemoryMap {
	size_t size;
	unsigned long magic;
	char data[];

 public:
	static uintptr_t create(size_t size) {
		size += sizeof(MemoryMap);
		if (auto mmap = Syscall::mmap(0, size, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0)) {
			MemoryMap * info = reinterpret_cast<MemoryMap *>(mmap.value());
			info->magic = MEMORYMAP_MAGIC;
			info->size = size;
			auto addr = reinterpret_cast<uintptr_t>(info->data);
			// Ensure GLIBC compatibility
			assert(addr % 16 == 0);
			return addr;
		} else {
			LOG_WARNING << "Allocator using mmap of " << size << " bytes failed: " << mmap.error_message() << endl;
			return 0;
		}
	}

	static uintptr_t resize(uintptr_t addr, size_t new_size) {
		new_size += sizeof(MemoryMap);
		uintptr_t start = addr - __builtin_offsetof(MemoryMap, data);
		MemoryMap * old_info = reinterpret_cast<MemoryMap *>(start);
		assert(old_info->magic == MEMORYMAP_MAGIC);
		if (auto mremap = Syscall::mremap(start, old_info->size, new_size, MREMAP_MAYMOVE)) {
			MemoryMap * new_info = reinterpret_cast<MemoryMap *>(mremap.value());
			assert(new_info->magic == MEMORYMAP_MAGIC);
			new_info->size = new_size;
			return reinterpret_cast<uintptr_t>(new_info->data);
		} else {
			LOG_WARNING << "Allocator using mremap of " << old_info->size << " bytes to " << new_size << " bytes at " << reinterpret_cast<void*>(start) << " failed: " << mremap.error_message() << endl;
			return 0;
		}
	}

	static void remove(uintptr_t addr) {
		uintptr_t start = addr - __builtin_offsetof(MemoryMap, data);
		MemoryMap * info = reinterpret_cast<MemoryMap *>(start);
		assert(info->magic == MEMORYMAP_MAGIC);
		size_t size = info->size;
		auto munmap = Syscall::munmap(start, size);
		if (munmap.failed()) {
			LOG_WARNING << "Allocator using munmap of " << size << " bytes at " << reinterpret_cast<void*>(start) << " failed: " << munmap.error_message() << endl;
		}
	}
};
#endif

uintptr_t alloc(size_t size) {
	if (size == 0) {
		return 0;
#ifndef DLH_LEGACY
	} else if (size >= MMAP_ALLOC_THREASHOLD) {
		return MemoryMap::create(size);
#endif
	} else {
		Guarded section(mutex);

		uintptr_t addr = allocator.malloc(size);

		// Ensure GLIBC compatibility
		assert(addr % 16 == 0);

		return addr;
	}
}

void free(uintptr_t addr) {
	if (addr != 0) {
#ifndef DLH_LEGACY
		if (allocator.size(addr) == 0) {
			MemoryMap::remove(addr);
		} else  // NOLINT
#endif
		 {
			Guarded section(mutex);

			allocator.free(addr);
		 }
	}
}

uintptr_t realloc(uintptr_t addr, size_t size) {
	if (addr == 0 && size == 0)
		return 0;

	size_t old_size = allocator.size(addr);
	uintptr_t new_addr = 0;

#ifndef DLH_LEGACY
	if ((addr != 0 && old_size == 0) || size >= MMAP_ALLOC_THREASHOLD) {
		if (size == 0) {
			MemoryMap::remove(addr);
		} else if (old_size == 0 && addr != 0) {
			new_addr = MemoryMap::resize(addr, size);
		} else {
			Guarded section(mutex);
			// First try resize in allocator (cheap)
			if (allocator.resize(addr, size)) {
				return addr;
			} else if ((new_addr = MemoryMap::create(size)) != 0 && addr != 0) {
				// Copy contents from allocator to mapping
				copy(new_addr, addr, Math::min(size, old_size));
				// Remove old memory
				allocator.free(addr);
			}
		}
	} else  // NOLINT
#endif
	 {
		assert(addr == 0 || old_size != 0);
		Guarded section(mutex);

		// Allocate (of size > 0)
		if (size > 0) {
			if (old_size != 0 && allocator.resize(addr, size))
				return addr;
			else if ((new_addr = allocator.malloc(size)) != 0 && addr != 0)
				// Copy contents
				copy(new_addr, addr, Math::min(size, old_size));
		}

		// Free old allocation
		if (addr != 0)
			allocator.free(addr);
	 }

	// Ensure GLIBC compatibility
	assert(new_addr % 16 == 0);
	return new_addr;
}

uintptr_t alloc_array(size_t nmemb, size_t size) {
	// Catch integer-multiplication overflow
	size_t bytes = nmemb * size;
	if (size == 0 || bytes / size != nmemb)
		return 0;

#ifndef DLH_LEGACY
	if (bytes >= MMAP_ALLOC_THREASHOLD)
		return MemoryMap::create(bytes);
#endif

	// Initialize memory
	uintptr_t new_addr = alloc(bytes);
	if (new_addr != 0)
		set(new_addr, 0, bytes);

	return new_addr;
}
}  // namespace Memory
