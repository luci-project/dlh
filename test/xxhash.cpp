#include <dlh/stream/output.hpp>
#include <dlh/xxhash.hpp>
#include <dlh/random.hpp>
#include <dlh/mem.hpp>

int main(int argc, const char *argv[]) {
	(void) argc;
	(void) argv;

	size_t n = 200;
	char * foo = Memory::alloc<char>(n);
	{
		Random random(23);
		for (size_t i = 0; i < n ; i++)
			foo[i] = static_cast<char>(random.number());
	}

	XXHash64 a(0);
	const char f[1337] = { 0 };
	a.add("foobar", 6);
	a.add(f, 33);
	a.add("ar rcs libdlh.a .build/utils/environ.o .build/utils/string.o .build/utils/mutex.o .build/utils/parser/string.o .build/utils/file.o .build/utils/auxiliary.o .build/utils/log.o .build/utils/stream/output.o .build/utils/stream/buffer.o .build/gnu/stack_chk.o .build/libc/stdarg.o .build/libc/stdlib.o .build/libc/string.o .build/libc/init.o .build/libc/errno.o .build/libc/assert.o .build/libc/unistd.o .build/libc/alloc.o .build/libc/stdio.o .build/libstdcxx/alloc.o .build/libstdcxx/atexit.o .build/libstdcxx/pure_virtual.o Makefile", 531);
	a.add(f, 1337);
	a.add("baz", 3);
	a.add(f, 1336);
	a.add("baz", 3);
	a.add(f, 1335);
	a.add("baz", 3);
	a.add(f, 1334);
	a.add("baz", 3);
	a.add(foo, 123);
	cout << "a: " << a.hash() << endl;

	XXHash64 b(0);
	b.add("foobar", 6);
	b.add(nullptr, 33);
	b.add("ar rcs libdlh.a .build/utils/environ.o .build/utils/string.o .build/utils/mutex.o .build/utils/parser/string.o .build/utils/file.o .build/utils/auxiliary.o .build/utils/log.o .build/utils/stream/output.o .build/utils/stream/buffer.o .build/gnu/stack_chk.o .build/libc/stdarg.o .build/libc/stdlib.o .build/libc/string.o .build/libc/init.o .build/libc/errno.o .build/libc/assert.o .build/libc/unistd.o .build/libc/alloc.o .build/libc/stdio.o .build/libstdcxx/alloc.o .build/libstdcxx/atexit.o .build/libstdcxx/pure_virtual.o Makefile", 531);
	b.addZeros(1337);
	b.add("baz", 3);
	b.addZeros(1336);
	b.add("baz", 3);
	b.addZeros(1335);
	b.add("baz", 3);
	b.addZeros(1334);
	b.add("baz", 3);
	b.add(foo, 123);
	cout << "b: " << b.hash() << endl;
	return 0;
}
