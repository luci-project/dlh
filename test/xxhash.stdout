a: 18245687198566367952
b: 18245687198566367952
